packages/app-form-builder/src/admin/components/FormEditor/Tabs/EditTab/FormStep/FormStep.tsx

```diff
@@ -0,0 +1,297 @@
+import { ReactComponent as EditIcon } from "@material-design-icons/svg/outlined/edit.svg";
+import { ReactComponent as HandleIcon } from "../../../../../icons/round-drag_indicator-24px.svg";
+
+import { Center, Vertical, Horizontal } from "../../../DropZone";
+
+import { i18n } from "@webiny/app/i18n";
+const t = i18n.namespace("FormsApp.Editor.EditTab");
+
+export const FormStep = ({
+    title,
+    deleteStepDisabled,
+    data,
+    formStep,
+    onDelete,
+    onEdit,
+    moveRow,
+    moveField,
+    getFieldPlugin,
+    insertField,
+    getLayoutFields,
+    updateField,
+    deleteField
+}: {
+    title: string;
+    deleteStepDisabled: boolean;
+    data: FbFormModel;
+    formStep: FbFormStep;
+    onDelete: () => void;
+    onEdit: () => void;
+    moveRow: (
+        source: number,
+        destination: number,
+        targetStepId: string,
+        sourceStep: FbFormStep
+    ) => void;
+    moveField: (params: MoveFieldParams) => void;
+    getFieldPlugin: ({ name }: { name: string }) => FbBuilderFieldPlugin | null;
+    insertField: (
+        field: FbFormModelField,
+        position: FieldLayoutPositionType,
+        stepId: string
+    ) => void;
+    getLayoutFields: (stepId: string) => FbFormModelField[][];
+    updateField: (field: FbFormModelField) => void;
+    deleteField: (field: FbFormModelField, stepId: string) => void;
+}) => {
+    const [editingField, setEditingField] = useState<FbFormModelField | null>(null);
+    const [dropTarget, setDropTarget] = useState<FieldLayoutPositionType | null>(null);
+
+    const editField = useCallback((field: FbFormModelField | null) => {
+        if (!field) {
+            setEditingField(null);
+            return;
+        }
+        setEditingField(cloneDeep(field));
+    }, []);
+
+    // TODO @ts-refactor figure out source type
+    const handleDropField = useCallback(
+        (source: any, position: FieldLayoutPositionType): void => {
+            const { pos, name, ui } = source;
+
+            if (name === "custom") {
+                /**
+                 * We can cast because field is empty in the start
+                 */
+                editField({} as FbFormModelField);
+                setDropTarget(position);
+                return;
+            }
+
+            if (ui === "row") {
+                // Reorder rows.
+                // Reorder logic is different depending on the source and target position.
+                moveRow(pos.row, position.row, formStep.id, pos.formStep);
+                return;
+            }
+
+            // If source pos is set, we are moving an existing field.
+            if (pos) {
+                if (pos.index === null) {
+                    console.log("Tried to move Form Field but its position index is null.");
+                    console.log(source);
+                    return;
+                }
+                const fieldId = (pos.formStep.id !== formStep.id ? pos.formStep : formStep).layout[
```

Comment:
Will add a comment that would explain what are we doing here

Reply:
Good point.